multicluster/apis/multicluster/v1alpha1/multiclusterset_types.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2021 Antrea Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+type ClusterStatus string
+
+const (
+	ClusterStatusReachable   ClusterStatus = "Reachable"
+	ClusterStatusUnreachable ClusterStatus = "Unreachable"
+)
+
+// MemberCluster defines member cluster information.
+type MemberCluster struct {
+	// Identify member cluster in multi-cluster set.
+	ClusterID string `json:"clusterID,omitempty"`
+	// API server of the member or the leader cluster.
+	Server string `json:"server,omitempty"`
+	// Secret name to access API server of the member from the leader cluster.
+	Secret string `json:"secret,omitempty"`
+	// ServiceAccount used by the member cluster to access into leader cluster.
+	ServiceAccount string `json:"serviceAccount,omitempty"`
+}
+
+// MemberClusterStatus defines member cluster status in the leader clusters.
+type MemberClusterStatus struct {
+	// ClusterID identifies member cluster in the set.
+	ClusterID string `json:"clusterID,omitempty"`
+	// IsLeader is true if the member cluster has selected this cluster as leader.
+	IsLeader bool `json:"isLeader,omitempty"`
+	// Status indicates connection status between leader and member cluster.
+	Status ClusterStatus `json:"status,omitempty"`
+	// Error reports of the member cluster.
+	Error string `json:"error,omitempty"`
+}
+
+// LeaderClusterStatus defines leader cluster status in the member clusters.
+type LeaderClusterStatus struct {
+	// ClusterID identifies leader cluster in the set.
+	ClusterID string `json:"clusterID,omitempty"`
+	// Status indicates connection status between leader and member cluster.
+	Status ClusterStatus `json:"status,omitempty"`
```

Comment:
can we implement Status as Conditions. See suggested api conventions for Status sub resource https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status

Reply:
of course.

Reply:
for ease of review, here is the proposed status for ClusterSet

```
type ClusterSetConditionType string
 
 
const (
    // Indicates whether ClusterSet is ready.
    ClusterSetReady         ClusterSetConditionType
)
 
 
type ClusterSetCondition struct {
    Type                    ClusterSetConditionType
    // Status of the condition, one of True, False, Unknown
    Status                  v1.ConditionStatus
 
 
    // +optional
    // Last time the condition transit from one status to another
    LastTransitionTime      metav1.Time
    // +optional
    // A human readable message indicating details about the transition
    Message                 string
    // +optional
    // Unique, one-word, CamelCase reason for the condition's last transition.
    Reason                  string
}
 
 
 
type ClusterConditionType string
 
const (
    // Indicates whether Cluster is ready and connected.
    ClusterReady            ClusterConditionType
    // Indicates whether Cluster is leader.
    ClusterIsLeader         ClusterConditionType
)
 
 
type ClusterCondition struct {
    Type                    ClusterConditionType
    // Status of the condition, one of True, False, Unknown
    Status                  v1.ConditionStatus
    // ClusterID is the unique identifier of this cluster.
    ClusterID               string
 
    // +optional
    // Last time the condition transit from one status to another
    LastTransitionTime      metav1.Time
    // +optional
    // A human readable message indicating details about the transition
    Message                 string
    // +optional
    // Unique, one-word, CamelCase reason for the condition's last transition.
    Reason                  string
}
 
 
type ClusterStatus struct {
    Conditions              []ClusterCondition
}
 
type ClusterSetStatus struct {
    // Total number of member clusters configured in the set.
    TotalClusters           int32
    // Total number of clusters ready and connected.
    ReadyClusters           int32
    // The overall condition of the cluster set.
    Conditions              []ClusterSetCondition
    // The status of individual member clusters.
    ClusterStatuses         []ClusterStatus
    // The generation observed by the controller.
    ObservedGeneration      int64
}
```